### PR TITLE
auth: add telemetry for why the user needed to reauth

### PR DESCRIPTION
**Pull request by nkomonen-amazon**
## Problem:
    
In the reauth metric (`aws_loginWithBrowser` + `isReAuth: true`) we didn't know why
the user needed to reauth. It could have been due to regular expiration or some unexpected
error.

## Solution:

**See each commit for specific change**

Add a new field in `aws_loginWithBrowser` named `reAuthReason` which will have some sort
of identifier for why reauth was needed.

What was actually changed in this PR:
- `SsoAccessTokenProvider` has an object `ReAuthState` (name may change) which maps an SSO connection -> reason why we need to reauth
  - In certain flows such as SSO token refresh, if it fails we will update the `ReAuthState` with the reason why.
  - Later on the user may want to reauthenticate, and when they start the flow `SsoAccessTokenProvider` will know the reason and record it in `aws_loginWithBrowser`
- `isReAuth` is now an argument in `createToken()`
  - This allows us to get rid of some hacky telemetry code where we tried to essentially set `isReAuth: true` globally (through `record()`)
  - Now that it is an argument we pass `isReAuth` throughout the execution, this is more readable. 
- A generalized Map class was created
  - This maps a key to some data
  - We used this implementation with `ReAuthState`
  - It can probably be reused in other places of our code and reduce duplication of code

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts
Comment (justinmk3) [inline packages/core/src/auth/sso/ssoAccessTokenProvider.ts:693]: is a class the right abstraction for this, or is "auth state" something that can/should be generalized in auth.ts?

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts
Comment (justinmk3) [inline packages/core/src/auth/sso/ssoAccessTokenProvider.ts:208]: is it possible this doesn't get called? then the "wrong" state may be used in the future

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts
Comment (justinmk3) [inline packages/core/src/auth/sso/ssoAccessTokenProvider.ts:703]: why use globalState instead of just a variable? what scenarios are we anticipating here?

File: packages/core/src/auth/auth.ts
Comment (justinmk3) [inline packages/core/src/auth/auth.ts:192]: this has the same risk as other global state, which e.g. codewhisperer has suffered from. 

are we certain this can't be achieved with a more direct plumbing-through of the "reauth reason", instead of setting global mutable state?

**Comment by justinmk3** [comment]
Thought experiement: in the places where this PR currently calls, `ReAuthReasonState.setReason`, why can't we use `telemetry.aws_loginWithBrowser.record({reAuthReason: ...})` instead?

If the answer is "because the execution context might be lost by the time we emit `telemetry.aws_loginWithBrowser`, then next questions are:

- can we change it so that the `telemetry.xx.record()` state that is built-up, is preserved until `telemetry.xx.run()` is called?
- or in the worst case, we could add a new function `telemetry.xx.recordNext()` which maintains state that is used for whenever the next `xx` metric is emitted.

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts
Comment (nkomonen-amazon) [reply packages/core/src/auth/sso/ssoAccessTokenProvider.ts:703]: If they are put in to a reauth state we do not know when they will reauth next. An IDE reload or extension update would wipe the state if it was in a variable.

Comment (justinmk3) [reply packages/core/src/auth/sso/ssoAccessTokenProvider.ts:703]: If the IDE reloads or restarts, is it actually valid to say that the "reauth reason" is something that happened minutes or hours ago?

Comment (nkomonen-amazon) [reply packages/core/src/auth/sso/ssoAccessTokenProvider.ts:703]: Yes, I would say so. It is the reason their connection was put in to a reauth state. Days or weeks later the cause is still the same, only once they reauth do we consider it stale.

Comment (justinmk3) [reply packages/core/src/auth/sso/ssoAccessTokenProvider.ts:703]: How common is the scenario that a vscode restart would happen after a reauth? It seems like an accident, and not connected to the reauth. Normally, the reauth will happen in the same session, so storing it in memory will capture enough data to be representative of the problems that we want to notice. 

If that's correct, then we don't need to store this in globalState, and that opens the door for using telemetry.record() as described in my other comments in this PR.

Comment (nkomonen-amazon) [reply packages/core/src/auth/sso/ssoAccessTokenProvider.ts:706]: If we are fine with assumption, then yes we can ignore storing it in globalState.

As for the implementation I will look in to something you mentioned like `telemetry.xx.recordNext()` as the telemetry context will be different and `record()` will not work as it currently is

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts
Comment (nkomonen-amazon) [inline packages/core/src/auth/sso/ssoAccessTokenProvider.ts:359]: We can inherit the parent constructor

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts
Comment (jpinkney-aws) [inline packages/core/src/auth/sso/ssoAccessTokenProvider.ts:75]: would it be possible to pass in a type that contains all the possible options? I think it might help to see all the possible invalidation reasons enumerated

Comment (nkomonen-amazon) [reply packages/core/src/auth/sso/ssoAccessTokenProvider.ts:75]: After looking at this again, some places prepend data to the string to give more context. Because of this it doesn't work nicely.

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts
Comment (justinmk3) [inline packages/core/src/auth/sso/ssoAccessTokenProvider.ts:673]: since ReAuthState lives as `this.reauthState` on the sso objects, can we eliminate this singleton to avoid accidental usage
```suggestion
```

File: packages/core/src/shared/utilities/map.ts
Comment (justinmk3) [inline packages/core/src/shared/utilities/map.ts:16]: most data is in-memory, so it seems confusing to call that out. if we add a non-memory variant then its name could call that out, since it would be less common.

```suggestion
export abstract class NestedMap<Key, Value = { [key: string]: any }> implements MapSync<Key, Value> {
```

File: packages/core/src/shared/utilities/map.ts
Comment (justinmk3) [inline packages/core/src/shared/utilities/map.ts:66]: What is this abstraction communicating? "not async" implies there is an async variant (where?).

Is there a reason this doesn't derive from the ES `Map` interface?

File: packages/core/src/shared/utilities/map.ts
Comment (justinmk3) [inline packages/core/src/shared/utilities/map.ts:47]: seems to overlap with `TelemetryDebounceInfo` https://github.com/aws/aws-toolkit-vscode/blob/d49eee404d1d65332c955c55d907116ae870e378/packages/core/src/shared/vscode/commands2.ts#L605

can we reconcile them? maybe we should start with the interface defined by `TelemetryDebounceInfo` and see what the delta is.

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts
Comment (nkomonen-amazon) [reply packages/core/src/auth/sso/ssoAccessTokenProvider.ts:673]: Nope, since the SsoAccessTokenProvider instance is created at the moment of use in different flows: https://github.com/aws/aws-toolkit-vscode/blob/b0965b7b768c957d740511ec2fd76a658794807a/packages/core/src/auth/auth.ts#L701.

We could move ReAuthState in to `auth.ts`, but I'd like to avoid that for now to keep ReAuthState contained to a single file.

File: packages/core/src/shared/utilities/map.ts
Comment (nkomonen-amazon) [reply packages/core/src/shared/utilities/map.ts:50]: This makes sense. I don't want it to block this PR anymore. I'll look to do this after we merge.

**Comment by nkomonen-amazon** [comment]
@justinmk3 can you take a final look when you have time

File: packages/core/src/shared/utilities/map.ts
Comment (justinmk3) [reply packages/core/src/shared/utilities/map.ts]: should this live in `cacheUtils.ts` ? it seems very related.

Comment (justinmk3) [reply packages/core/src/shared/utilities/map.ts]: > There are some differences with the Map interface, 

is that difference big enough to lose the discoverability/intuitiveness gained by implementing `Map` ? 

> such as not allowing partial values in `set()` that wouldn't allow it to be used.

I am wondering about this. Is it just a typing issue? I think we can work around it.

File: packages/core/src/shared/utilities/cacheUtils.ts
Comment (justinmk3) [inline packages/core/src/shared/utilities/cacheUtils.ts:22]: what indicates that `KeyedCache` doesn't have atomicity? can be a followup discussion, but... this isn't clear.

File: packages/core/src/shared/utilities/map.ts
Comment (justinmk3) [inline packages/core/src/shared/utilities/map.ts:14]: (followup) this is called "hash" usually, isn't it? we can look for ways to give clues to other developers, by using common terms. 

```suggestion
 * - A key does not have to be a string, {@link NestedMap.hash} enables this
```

File: packages/core/src/test/credentials/auth.test.ts
Comment (justinmk3) [inline packages/core/src/test/credentials/auth.test.ts:262]: can't we test the actual metric, as before?

File: packages/core/src/test/credentials/sso/ssoAccessTokenProvider.test.ts
Comment (justinmk3) [inline packages/core/src/test/credentials/sso/ssoAccessTokenProvider.test.ts:245]: we have existing tests for the various auth cases. can we just add asserts about reAuth in those existing tests? reauth itself is not a topic, it's just a field. 

It helps connect concepts if auth fields are all tested in the auth tests, instead of having separate test groups for every piece of auth state.

**Review by justinmk3 — APPROVED** [review]
approving to include in today's release, comments can be a fast-follow

File: packages/core/src/test/credentials/auth.test.ts
Comment (nkomonen-amazon) [reply packages/core/src/test/credentials/auth.test.ts:262]: This suite already stubs the underlying sso token provider `createToken()` in the test so no telemetry calls are actually made in this case.

This test only existed due to the `aws_loginWithBrowser` hack we previously used in `auth.ts`. Now that all telemetry is isolated to `SsoAccessTokenProvider.ts` telemetry is asserted in the SsoTokenProvider tests.

File: packages/core/src/test/credentials/sso/ssoAccessTokenProvider.test.ts
Comment (nkomonen-amazon) [reply packages/core/src/test/credentials/sso/ssoAccessTokenProvider.test.ts:245]: Updated

File: packages/core/src/shared/utilities/cacheUtils.ts
Comment (nkomonen-amazon) [reply packages/core/src/shared/utilities/cacheUtils.ts:22]: Making updates requires async read() then write(). So by the time you write the data may have changed due to async. But since MapSync does not use promises the user can read the latest data, then update it atomically.

**Pull request merged**
